### ISSUE TITLE
Refreshing state throws "Error: Invalid for_each argument" because local.azurerm_policy_definition_external_lookup will be known only after apply

### ISSUE THREAD

**Issue by benkoben** [Azure/terraform-azurerm-caf-enterprise-scale#511]
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Versions

**terraform**: 1.3.4 (also tried with 1.3.2)

**azure provider**: ">=3.2"

**module**:  Azure/caf-enterprise-scale/azurerm : 2.3.0

### Description

Refreshing state throws "Error: Invalid for_each argument". This started occuring after I changed the displayName of a pre-existing assignment. Full error:

```
launcher0_1  | module.enterprise_scale.time_sleep.after_azurerm_policy_assignment: Refreshing state... [id=2022-11-04T10:44:46Z]
launcher0_1  | 
launcher0_1  |  Error: Invalid for_each argument
launcher0_1  |  
launcher0_1  |    on .terraform/modules/enterprise_scale/locals.policy_assignments.tf line 161, in data "azurerm_policy_definition" "external_lookup":
launcher0_1  |   161:   for_each = local.azurerm_policy_definition_external_lookup
launcher0_1  |      
launcher0_1  |       local.azurerm_policy_definition_external_lookup will be known only after apply
launcher0_1  |  
launcher0_1  |  The "for_each" map includes keys derived from resource attributes that
launcher0_1  |  cannot be determined until apply, and so Terraform cannot determine the
launcher0_1  |  full set of keys that will identify the instances of this resource.
launcher0_1  |  
launcher0_1  |  When working with unknown values in for_each, it's better to define the map
launcher0_1  |  keys statically in your configuration and place apply-time results only in
launcher0_1  |  the map values.
launcher0_1  |  
launcher0_1  |  Alternatively, you could use the -target planning option to first apply
launcher0_1  |  only the resources that the for_each value depends on, and then apply a
launcher0_1  |  second time to fully converge.
```
From my understanding the module tried to perform a lookup of external definitions which are referenced in policy assignments that have an identity type set to SystemAssigned. These definitions are saved to **local.azurerm_policy_definition_external_lookup**. 

The assignment which I was working on when this error occured references a custom initiative which in turn uses the following two built-in definitions:
* /providers/Microsoft.Authorization/policyDefinitions/ea3f2387-9b95-492a-a190-fcdc54f7b070
* /providers/Microsoft.Authorization/policyDefinitions/40df99da-1232-49b1-a39a-6da8d878f469

These definitionIds are hardcoded into the initiative templates, and then the initiative is hardcoded into the assignment template, which makes me wonder how "a values only known after apply" has found its way into **local.azurerm_policy_definition_external_lookup**.

I have attached example files in this issue. These are not exact replicas but instead simplified for obfuscation purposes.

Steps that I have tried:

* Running plan from an older commit
* Removing all resource addreses related to my changes from the state
* Made sure to use latest version of terraform-azurerm-caf-enterprise
* I also added the following two outputs to the downloaded module
```
# Debugging purposes
output "external_lookup_definitions" {
  value = local.azurerm_policy_definition_external_lookup
  description = "I use this for debugging"
}

# Debugging purposes
output "external_lookup_definitions_data" {
  value = data.azurerm_policy_definition.external_lookup
  description = "I use this for debugging 2"
}
```

Accessing these outputs via the terraform console works perfectly fine.

#### Describe the bug

**local.azurerm_policy_definition_external_lookup** is populated with a dynamic value causing Error: Invalid for_each argument

Not sure how I should continue troubleshooting because it is unclear to me if this is a bug or not.

#### Steps to Reproduce

1. Create the policySetDefinitions I have attached to this issue. Modify the MANAGEMENT_GROUP_NAME in the assignment templates before applying.
2. Create the policyAssignment I have attached to this issue. Modify the MANAGEMENT_GROUP_NAME in the assignment templates before applying.
3. Apply
4. Change the displayName of the assignment and perform a new plan/apply

#### Screenshots

Change I made before error occured
![image](https://user-images.githubusercontent.com/100195282/201363199-07b5915a-4c5b-401d-b9e7-58b215d04b6c.png)

[modules.zip](https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/files/9990980/modules.zip)
